exporter/ststopologyexporter/topology.go

```diff
@@ -0,0 +1,170 @@
+// Copyright StackState B.V.
+// SPDX-License-Identifier: Apache-2.0
+package ststopologyexporter
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/zap"
+)
+
+var (
+	errInternal = errors.New("internal error")
+)
+
+type topologyExporter struct {
+	logger     *zap.Logger
+	httpClient http.Client
+	cfg        *Config
+}
+
+func newTopologyExporter(logger *zap.Logger, cfg component.Config) (*topologyExporter, error) {
+	stsCfg, ok := cfg.(*Config)
+	if !ok {
+		return nil, fmt.Errorf("invalid config passed to stackstateexporter: %T", cfg)
+	}
+	httpClient := http.Client{
+		Timeout: 5 * time.Second, // TODO configure timeout
+	}
+
+	return &topologyExporter{logger: logger, httpClient: httpClient, cfg: stsCfg}, nil
+}
+
+type Instance struct {
+	Type string `json:"type"`
+	URL  string `json:"url"`
+}
+
+type ComponentType struct {
+	Name string `json:"name"`
+}
+
+type Component struct {
+	ExternalId string            `json:"externalId"`
+	Type       ComponentType     `json:"type"`
+	Data       map[string]string `json:"data"`
+}
+
+type Relation struct{}
+
+type Topology struct {
+	Instance   Instance    `json:"instance"`
+	Components []Component `json:"components"`
+	Relations  []Relation  `json:"relations"`
+}
+
+type IntakeTopology struct {
+	CollectionTimestamp int64      `json:"collection_timestamp"`
+	InternalHostname    string     `json:"internalHostname"`
+	Topologies          []Topology `json:"topologies"`
+}
+
+func (t *topologyExporter) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) error {
+	log := t.logger
+
+	componentsByApiKey := make(map[string][]Component, 0)
+	rms := md.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		rs := rms.At(i)
+		resource := rs.Resource()
+		serviceName, ok := resource.Attributes().Get("service.name")
+		if !ok {
+			log.Warn("Skipping resource without service.name")
+			continue
+		}
+		serviceNamespace, ok := resource.Attributes().Get("service.namespace")
+		if !ok {
+			log.Warn("Skipping resource without service.namespace")
+			continue
+		}
+		instanceId, ok := resource.Attributes().Get("service.instance.id")
+		var serviceInstanceId pcommon.Value
+		if !ok {
+			serviceInstanceId = serviceName
+		} else {
+			serviceInstanceId = instanceId
+		}
+		sts_api_key, key_exists := resource.Attributes().Get("sts_api_key")
+		if !key_exists {
+			log.Warn("No sts_api_key attribute found on resource")
+			continue
+		}
+		components, has_siblings := componentsByApiKey[sts_api_key.AsString()]
+		if !has_siblings {
+			components = make([]Component, 0)
+		}
+		components = append(components, Component{
+			fmt.Sprintf("%s-%s", serviceNamespace.AsString(), serviceName.AsString()),
```

Comment:
The common approach for topology is that the external data source provides the data ready for usage with minimal required processing in StackState. So I suggest we do that here as well. 

For the externalId it means that it already is in the form of a `urn` that stackstate can use as an identifier. I'm thinking of  `fmt.Sprintf("urn:opentelemetry:%s:service/%s", serviceNamespace.AsString(), serviceName.AsString()),` for services and similar `fmt.Sprintf("urn:opentelemetry:%s:service/%s:serviceInstance/%s", serviceNamespace.AsString(), serviceName.AsString(), serviceInstanceId.AsString())` for service instance.